wolfcrypt/src/hpke.c

```diff
@@ -0,0 +1,1139 @@
+static const int EAE_PRK_LABEL_STR_LEN = 7;
+
+static const char* SHARED_SECRET_LABEL_STR = "shared_secret";
+static const int SHARED_SECRET_LABEL_STR_LEN = 13;
+
+static const char* PSK_ID_HASH_LABEL_STR = "psk_id_hash";
+static const int PSK_ID_HASH_LABEL_STR_LEN = 11;
+
+static const char* INFO_HASH_LABEL_STR = "info_hash";
+static const int INFO_HASH_LABEL_STR_LEN = 9;
+
+static const char* SECRET_LABEL_STR = "secret";
+static const int SECRET_LABEL_STR_LEN = 6;
+
+static const char* KEY_LABEL_STR = "key";
+static const int KEY_LABEL_STR_LEN = 3;
+
+static const char* BASE_NONCE_LABEL_STR = "base_nonce";
+static const int BASE_NONCE_LABEL_STR_LEN = 10;
+
+static const char* EXP_LABEL_STR = "exp";
+static const int EXP_LABEL_STR_LEN = 3;
+
+/* encode n as a byte string with length w, return 0 or error */
+static int I2OSP(int n, int w, byte* out)
+{
+    int i;
+
+    if (w <= 0 || w > 32) {
+        return MP_VAL;
+    }
+
+    /* if width is less than int max check that n is less than w bytes max */
+    /* if width is greater than int max check that n is less than int max */
+    if ((w < 4 && n > ((1 << (w * 8)) - 1)) || (w >= 4 && n > 0x7fffffff)) {
+        return MP_VAL;
+    }
+
+    /* make sure the byte string is cleared */
+    XMEMSET( out, 0, w );
+
+    for (i = 0; i < w && n > 0; i++) {
+        out[w-(i + 1)] = (byte)n;
+        n >>= 8;
+    }
+
+    return 0;
+}
+
+int wc_HpkeInit(Hpke* hpke, int kem, int kdf, int aead, void* heap)
+{
+    int ret;
+    byte* id;
+
+    if (hpke == NULL || kem == 0 || kdf == 0 || aead == 0) {
+        return BAD_FUNC_ARG;
+    }
+
+    XMEMSET(hpke, 0, sizeof(*hpke));
+    hpke->kem = kem;
+    hpke->kdf = kdf;
+    hpke->aead = aead;
+    hpke->heap = heap;
+
+    /* set kem_suite_id */
+    id = hpke->kem_suite_id;
+
+    XMEMCPY(id, KEM_STR, KEM_STR_LEN);
+    id += KEM_STR_LEN;
+
+    ret = I2OSP(kem, 2, id);
+
+    /* set hpke_suite_id */
+    id = hpke->hpke_suite_id;
+
+    XMEMCPY(id, HPKE_STR, HPKE_STR_LEN);
+    id += HPKE_STR_LEN;
+
+    if (ret == 0) {
+      ret = I2OSP(kem, 2, id);
+      id += 2;
+    }
+    if (ret == 0) {
+      ret = I2OSP(kdf, 2, id);
+      id += 2;
+    }
+    if (ret == 0) {
+      ret = I2OSP(aead, 2, id);
+    }
+    if (ret != 0)
+        return ret;
+
+    switch (kem)
+    {
+#if defined(HAVE_ECC)
+#if defined(WOLFSSL_SHA224) || !defined(NO_SHA256)
+        case DHKEM_P256_HKDF_SHA256:
+            hpke->curve_id = ECC_SECP256R1;
+            hpke->Nsecret = WC_SHA256_DIGEST_SIZE;
+            hpke->Nh = WC_SHA256_DIGEST_SIZE;
+            hpke->Ndh = wc_ecc_get_curve_size_from_id( hpke->curve_id );
+            hpke->Npk = 1 + hpke->Ndh * 2;
+            break;
+#endif
+
+#ifdef WOLFSSL_SHA384
+        case DHKEM_P384_HKDF_SHA384:
+            hpke->curve_id = ECC_SECP384R1;
+            hpke->Nsecret = WC_SHA384_DIGEST_SIZE;
+            hpke->Nh = WC_SHA384_DIGEST_SIZE;
+            hpke->Ndh = wc_ecc_get_curve_size_from_id( hpke->curve_id );
```

Comment:
wc_ecc_get_curve_size_from_id can return a negative value on failure, we should check that Ndh here and later on in this function is not getting set as an error value